File: forge-core/internal/governancecontract/record_set.go
package governancecontract

import (
	"bytes"
	"fmt"
	"strconv"
)

func DecodeRecordSet(data []byte) ([]*Record, error) {
	node, err := parseStrictJSONBounded(data, maxSetBytes)
	if err != nil {
		return nil, fmt.Errorf("governance record set JSON: %w", err)
	}
	array, ok := node.([]any)
	if !ok || len(array) == 0 {
		return nil, fmt.Errorf("governance record set must be a nonempty array")
	}
	canonical, err := canonicalJSON(array)
	if err != nil {
		return nil, err
	}
	if !bytes.Equal(data, canonical) {
		return nil, fmt.Errorf("record set input is not exact compact canonical JSON")
	}
	records, err := decodeRecordNodes(array)
	if err != nil {
		return nil, err
	}
	if err := ValidateRecordSet(records); err != nil {
		return nil, err
	}
	return records, nil
}

func decodeRecordNodes(nodes []any) ([]*Record, error) {
	records := make([]*Record, 0, len(nodes))
	for index, node := range nodes {
		encoded, err := canonicalJSON(node)
		if err != nil {
			return nil, fmt.Errorf("record %d: %w", index, err)
		}
		if len(encoded) > maxRecordBytes {
			return nil, fmt.Errorf("record %d exceeds %d bytes", index, maxRecordBytes)
		}
		record, err := DecodeRecord(encoded)
		if err != nil {
			return nil, fmt.Errorf("record %d: %w", index, err)
		}
		records = append(records, record)
	}
	return records, nil
}

func ValidateRecordSet(records []*Record) error {
	if len(records) == 0 || len(records) > maxArrayItems {
		return fmt.Errorf("record set must contain 1..%d records", maxArrayItems)
	}
	byID, err := indexRecords(records)
	if err != nil {
		return err
	}
	if err := detectSupersessionCycles(records, byID); err != nil {
		return err
	}
	if err := detectClaimDerivationCycles(records, byID); err != nil {
		return err
	}
	for _, record := range records {
		if err := validateSupersession(record, byID); err != nil {
			return err
		}
		if record.Claim != nil {
			if err := validateClaimReferences(record.Claim, byID); err != nil {
				return err
			}
		}
	}
	return nil
}

func indexRecords(records []*Record) (map[string]*Record, error) {
	byID := make(map[string]*Record, len(records))
	tuples := make(map[string]bool, len(records))
	totalBytes, previousID := len(records)+1, ""
	for index, record := range records {
		recordBytes, err := validateSetRecord(record)
		if err != nil {
			return nil, fmt.Errorf("record %d: %w", index, err)
		}
		header := record.Header()
		if index > 0 && header.RecordID <= previousID {
			return nil, fmt.Errorf("record set must be sorted by unique record_id")
		}
		if _, exists := byID[header.RecordID]; exists {
			return nil, fmt.Errorf("duplicate record_id %q", header.RecordID)
		}
		tuple := record.Kind() + "\x00" + header.AggregateID + "\x00" + strconv.FormatInt(header.Sequence, 10)
		if tuples[tuple] {
			return nil, fmt.Errorf("duplicate kind/aggregate_id/sequence tuple")
		}
		byID[header.RecordID], tuples[tuple] = record, true
		totalBytes, previousID = totalBytes+recordBytes, header.RecordID
	}
	if totalBytes > maxSetBytes {
		return nil, fmt.Errorf("record set exceeds %d canonical bytes", maxSetBytes)
	}
	return byID, nil
}

func validateSetRecord(record *Record) (int, error) {
	state, err := verifiedCanonicalState(record)
	if err != nil {
		return 0, err
	}
	applyCanonicalState(record, state)
	return len(state.record), nil
}

func validateSupersession(record *Record, byID map[string]*Record) error {
	header := record.Header()
	if header.Sequence == 1 && len(header.SupersedesRecordIDs) != 0 {
		return fmt.Errorf("sequence 1 record %q cannot supersede records", header.RecordID)
	}
	if header.Sequence > 1 && len(header.SupersedesRecordIDs) == 0 {
		return fmt.Errorf("sequence %d record %q must supersede records", header.Sequence, header.RecordID)
	}
	hasPriorSequence := false
	for _, priorID := range header.SupersedesRecordIDs {
		prior, exists := byID[priorID]
		if !exists {
			return fmt.Errorf("record %q supersedes missing record %q", header.RecordID, priorID)
		}
		priorHeader := prior.Header()
		if prior.Kind() != record.Kind() || priorHeader.AggregateID != header.AggregateID || priorHeader.Sequence >= header.Sequence {
			return fmt.Errorf("record %q has invalid supersession target %q", header.RecordID, priorID)
		}
		hasPriorSequence = hasPriorSequence || priorHeader.Sequence == header.Sequence-1
	}
	if header.Sequence > 1 && !hasPriorSequence {
		return fmt.Errorf("record %q must supersede a sequence %d record", header.RecordID, header.Sequence-1)
	}
	return nil
}

func detectSupersessionCycles(records []*Record, byID map[string]*Record) error {
	visiting, visited := make(map[string]bool), make(map[string]bool)
	var visit func(string) error
	visit = func(recordID string) error {
		if visiting[recordID] {
			return fmt.Errorf("supersession cycle includes record %q", recordID)
		}
		if visited[recordID] {
			return nil
		}
		visiting[recordID] = true
		for _, priorID := range byID[recordID].Header().SupersedesRecordIDs {
			if _, exists := byID[priorID]; exists {
				if err := visit(priorID); err != nil {
					return err
				}
			}
		}
		delete(visiting, recordID)
		visited[recordID] = true
		return nil
	}
	for _, record := range records {
		if err := visit(record.Header().RecordID); err != nil {
			return err
		}
	}
	return nil
}

func validateClaimReferences(claim *KnowledgeClaim, byID map[string]*Record) error {
	for _, recordID := range append(append([]string{}, claim.Spec.SupportingEvidenceRecordIDs...), claim.Spec.ContradictingEvidenceRecordIDs...) {
		referenced, exists := byID[recordID]
		if !exists || referenced.Evidence == nil {
			return fmt.Errorf("claim %q references missing EvidenceRecord %q", claim.Metadata.RecordID, recordID)
		}
		if !containsString(referenced.Evidence.Spec.Subjects, claim.Spec.Subject) {
			return fmt.Errorf("EvidenceRecord %q does not contain claim subject %q", recordID, claim.Spec.Subject)
		}
	}
	for _, recordID := range claim.Spec.DerivedFromClaimRecordIDs {
		referenced, exists := byID[recordID]
		if !exists || referenced.Claim == nil {
			return fmt.Errorf("claim %q references missing KnowledgeClaim %q", claim.Metadata.RecordID, recordID)
		}
	}
	return nil
}

func detectClaimDerivationCycles(records []*Record, byID map[string]*Record) error {
	visiting, visited := make(map[string]bool), make(map[string]bool)
	var visit func(string) error
	visit = func(recordID string) error {
		if visiting[recordID] {
			return fmt.Errorf("claim derivation cycle includes record %q", recordID)
		}
		if visited[recordID] {
			return nil
		}
		visiting[recordID] = true
		for _, derivedID := range byID[recordID].Claim.Spec.DerivedFromClaimRecordIDs {
			if derivedID == recordID {
				return fmt.Errorf("claim %q cannot derive from itself", recordID)
			}
			if target, exists := byID[derivedID]; exists && target.Claim != nil {
				if err := visit(derivedID); err != nil {
					return err
				}
			}
		}
		delete(visiting, recordID)
		visited[recordID] = true
		return nil
	}
	for _, record := range records {
		if record.Claim != nil {
			if err := visit(record.Header().RecordID); err != nil {
				return err
			}
		}
	}
	return nil
}

func containsString(values []string, target string) bool {
	for _, value := range values {
		if value == target {
			return true
		}
	}
	return false
}

File: harness/governance_contract/record_set.py
"""Record-set identity, reference-closure and supersession validation."""

from __future__ import annotations

from .codec import ContractError, canonical_json, decode_record_set
from .constants import MAX_ITEMS, MAX_SET_BYTES
from .semantics import validate_record


def _record_index(records: list[dict[str, object]], issues: list[str]) -> dict[str, dict[str, object]]:
    by_id: dict[str, dict[str, object]] = {}
    identities: set[tuple[object, object, object]] = set()
    record_ids = []
    for record in records:
        metadata = record.get("metadata")
        record_ids.append(metadata.get("record_id") if isinstance(metadata, dict) else None)
    if not all(isinstance(item, str) for item in record_ids) or record_ids != sorted(record_ids):
        issues.append("record set must be sorted by record_id")
    for index, record in enumerate(records):
        metadata = record.get("metadata")
        if not isinstance(metadata, dict):
            continue
        record_id = metadata.get("record_id")
        identity = (record.get("kind"), metadata.get("aggregate_id"), metadata.get("sequence"))
        if isinstance(record_id, str) and record_id in by_id:
            issues.append(f"records[{index}]: duplicate record_id {record_id!r}")
        elif isinstance(record_id, str):
            by_id[record_id] = record
        if (isinstance(identity[0], str) and isinstance(identity[1], str) and
                type(identity[2]) is int):
            if identity in identities:
                issues.append(f"records[{index}]: duplicate kind/aggregate_id/sequence")
            identities.add(identity)
    return by_id


def _validate_supersession(record: dict[str, object], by_id: dict[str, dict[str, object]],
                           label: str, issues: list[str]) -> None:
    metadata = record.get("metadata")
    if not isinstance(metadata, dict) or not isinstance(metadata.get("supersedes_record_ids"), list):
        return
    prior_ids, sequence = metadata["supersedes_record_ids"], metadata.get("sequence")
    if sequence == 1 and prior_ids:
        issues.append(f"{label}: sequence 1 cannot supersede records")
    if type(sequence) is int and sequence > 1 and not prior_ids:
        issues.append(f"{label}: sequence > 1 must supersede an exact prior record")
    prior_sequences = set()
    for prior_id in prior_ids:
        if not isinstance(prior_id, str):
            continue
        prior = by_id.get(prior_id)
        prior_meta = prior.get("metadata") if isinstance(prior, dict) else None
        if not isinstance(prior_meta, dict):
            issues.append(f"{label}: supersedes unknown record {prior_id!r}")
        elif (prior.get("kind") != record.get("kind") or
              prior_meta.get("aggregate_id") != metadata.get("aggregate_id") or
              not (type(prior_meta.get("sequence")) is int and type(sequence) is int and
                   prior_meta["sequence"] < sequence)):
            issues.append(f"{label}: superseded record must be same kind/aggregate and lower sequence")
        else:
            prior_sequences.add(prior_meta["sequence"])
    if type(sequence) is int and sequence > 1 and sequence - 1 not in prior_sequences:
        issues.append(f"{label}: supersedes must include a sequence-1 record")


def _validate_claim_refs(record: dict[str, object], by_id: dict[str, dict[str, object]],
                         label: str, issues: list[str]) -> None:
    if record.get("kind") != "KnowledgeClaim" or not isinstance(record.get("spec"), dict):
        return
    spec, subject = record["spec"], record["spec"].get("subject")
    evidence_fields = {"supporting_evidence_record_ids", "contradicting_evidence_record_ids"}
    for field in evidence_fields:
        refs = spec.get(field)
        if not isinstance(refs, list):
            continue
        for reference in refs:
            if not isinstance(reference, str):
                continue
            target = by_id.get(reference)
            target_spec = target.get("spec") if isinstance(target, dict) else None
            subjects = target_spec.get("subjects") if isinstance(target_spec, dict) else None
            if not isinstance(target, dict) or target.get("kind") != "EvidenceRecord":
                issues.append(f"{label}.spec.{field}: unknown EvidenceRecord {reference!r}")
            elif not isinstance(subjects, list) or subject not in subjects:
                issues.append(f"{label}.spec.{field}: evidence does not cover claim subject")
    derived = spec.get("derived_from_claim_record_ids")
    if not isinstance(derived, list):
        return
    for reference in derived:
        if not isinstance(reference, str):
            continue
        target = by_id.get(reference)
        if not isinstance(target, dict) or target.get("kind") != "KnowledgeClaim":
            issues.append(f"{label}.spec.derived_from_claim_record_ids: unknown KnowledgeClaim {reference!r}")


def _supersession_cycle_issues(records: list[dict[str, object]], issues: list[str]) -> None:
    graph = {}
    for record in records:
        metadata = record.get("metadata")
        if isinstance(metadata, dict) and isinstance(metadata.get("record_id"), str):
            refs = metadata.get("supersedes_record_ids")
            graph[metadata["record_id"]] = refs if isinstance(refs, list) else []
    visiting: set[str] = set()
    visited: set[str] = set()

    def visit(node: str) -> bool:
        if node in visiting:
            return True
        if node in visited:
            return False
        visiting.add(node)
        cyclic = any(isinstance(prior, str) and prior in graph and visit(prior)
                     for prior in graph.get(node, []))
        visiting.remove(node)
        visited.add(node)
        return cyclic

    if any(visit(node) for node in graph if node not in visited):
        issues.append("record set contains a supersession cycle")


def _derivation_cycle_issues(records: list[dict[str, object]], issues: list[str]) -> None:
    graph = {}
    for record in records:
        metadata, spec = record.get("metadata"), record.get("spec")
        if (record.get("kind") == "KnowledgeClaim" and isinstance(metadata, dict) and
                isinstance(spec, dict) and isinstance(metadata.get("record_id"), str)):
            refs = spec.get("derived_from_claim_record_ids")
            graph[metadata["record_id"]] = refs if isinstance(refs, list) else []
    visiting: set[str] = set()
    visited: set[str] = set()

    def visit(node: str) -> bool:
        if node in visiting:
            return True
        if node in visited:
            return False
        visiting.add(node)
        cyclic = any(isinstance(prior, str) and prior in graph and visit(prior)
                     for prior in graph.get(node, []))
        visiting.remove(node)
        visited.add(node)
        return cyclic

    if any(visit(node) for node in graph if node not in visited):
        issues.append("record set contains a claim derivation cycle")


def _validate_record_set(records: list[dict[str, object]]) -> list[str]:
    if not isinstance(records, list) or not records or len(records) > MAX_ITEMS:
        return [f"record set must contain 1..{MAX_ITEMS} records"]
    try:
        if len(canonical_json(records)) > MAX_SET_BYTES:
            return [f"record set exceeds {MAX_SET_BYTES} bytes"]
    except ContractError as error:
        return [str(error)]
    issues: list[str] = []
    for index, record in enumerate(records):
        if not isinstance(record, dict):
            issues.append(f"records[{index}]: expected object")
        else:
            validate_record(record, index, issues)
    valid_records = [record for record in records if isinstance(record, dict)]
    by_id = _record_index(valid_records, issues)
    for index, record in enumerate(records):
        if isinstance(record, dict):
            _validate_supersession(record, by_id, f"records[{index}]", issues)
            _validate_claim_refs(record, by_id, f"records[{index}]", issues)
    _supersession_cycle_issues(valid_records, issues)
    _derivation_cycle_issues(valid_records, issues)
    return issues


def validate_record_set(records: list[dict[str, object]]) -> list[str]:
    """Return bounded issues without granting truth, authority or completion."""
    try:
        return _validate_record_set(records)
    except MemoryError:
        return ["record-set validation exhausted memory"]


def check_record_set_bytes(raw: bytes) -> list[str]:
    """Decode and validate canonical record-set bytes, returning bounded issues."""
    try:
        return validate_record_set(decode_record_set(raw))
    except ContractError as error:
        return [str(error)]
    except MemoryError:
        return ["record-set processing exhausted memory"]
